### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by ljabdo**
Please review so we can merge unit tests into integ.

### CODE REVIEW

**Comment by Darrick-Oliver** [comment]
Looks like the formatting test failed, you may want to try re-running `npm run format`

File: backend/tests/MessageController.test.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { MessageController } from '../controllers/MessageController';
+import { bootstrapDB } from '../bootstrap/db_init';
+import mongoose from 'mongoose';
+import { CurrentUser } from 'routing-controllers';
+
+const mc = new MessageController();
+jest.setTimeout(10000);
+
+beforeAll(async () => {
+    await mongoose.disconnect();
```
Comment (Darrick-Oliver) [inline backend/tests/MessageController.test.ts:10]: Not a blocker but why are we disconnecting before we connect?

File: backend/tests/MessageController.test.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    await bootstrapDB();
+});
+
+afterAll(async () => {
+    await mongoose.disconnect();
+    console.log('MongDB disconnected');
+});
+
+test('Test - Send Message - fail', async () => {
```
Comment (Darrick-Oliver) [inline backend/tests/MessageController.test.ts:19]: I'm assuming these aren't complete yet, right?

File: backend/tests/UserController.test.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    await mongoose.disconnect();
+    await bootstrapDB();
+});
+
+afterAll(async () => {
+    await mongoose.disconnect();
+    console.log('MongDB disconnected');
+});
+
+test('userController - login', async () => {
```
Comment (Darrick-Oliver) [inline backend/tests/UserController.test.ts:18]: We may want to revisit these tests in the future using the local db that the professor suggested. It works now, but its unreliable.

**Review by Darrick-Oliver — COMMENTED** [review]
Looks 90% of the way there, thanks for making all of these tests. You have pretty decent coverage so far.

File: backend/package.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
     "description": "JibJab backend",
     "main": "index.js",
     "scripts": {
-        "test": "prettier --check .",
+        "test": "jest --detectOpenHandles --collectCoverage",
```
[the remark this answers is not in the document]
Comment (ljabdo) [reply backend/package.json:7]: Sounds good I will fix

**Comment by ljabdo** [comment]
Will change testing suite to use in memory database thanks for the feedback team

File: backend/coverage/clover.xml [hunk 5]
```diff
@@ -0,0 +1,108 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Darrick-Oliver) [inline backend/coverage/clover.xml:1]: I like that you added coverage checks. Maybe we can move the actual reports to a gitignore though, not sure if its too much clutter. Would like some comments from others with opinions too

File: backend/package.json [hunk 6]
```diff
@@ -18,6 +18,7 @@
         "express": "^4.18.2",
         "jest": "^29.2.1",
         "jsonwebtoken": "^8.5.1",
+        "mongodb-memory-server": "^8.10.1",
```
Comment (Darrick-Oliver) [inline backend/package.json:21]: 👏👏🥳

**Review by soderdrew — APPROVED** [review]
I like the descriptiveness and consistent format of the test names. Good work!

File: backend/tests/UserController.test.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+afterAll(async () => {
+    //await mongoose.disconnect();
+    //console.log('MongDB disconnected');
+    await mongoose.disconnect();
+    if (mongod){
+        await mongod.stop();
+    }
+});
+
+test('simple test', async () => {
```
Comment (Darrick-Oliver) [inline backend/tests/UserController.test.ts:31]: Is this just a sanity check?

File: backend/tests/UserController.test.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    try {
+        const result = await uc.register(
+            username,
+            first_name,
+            last_name,
+            email,
+            password
+        );
+    } catch (err: any) {
+        expect(err.error).toBe(true);
```
Comment (Darrick-Oliver) [inline backend/tests/UserController.test.ts:54]: Might want to check return message as well

File: backend/tests/UserController.test.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    }
+});
+
+test('userController - login', async () => {
+    const email = 'helloFresh';
+    const password = 'thisPasswordWorks';
+    const result = await uc.login(email, password);
+    expect(result.error).toBe(false);
```
Comment (Darrick-Oliver) [inline backend/tests/UserController.test.ts:135]: I love the tests youve got so far -- just maybe check the resulting message too just to be sure we're catching the right error

**Review by Darrick-Oliver — APPROVED** [review]

File: backend/tests/UserController.test.ts (hunk 8 above)
Comment (Darrick-Oliver) [reply backend/tests/UserController.test.ts:54]: Same callout to all the below tests too

**Pull request merged**